### PR TITLE
Add directory-partitioned data support to cudf.read_parquet

### LINKED ISSUE #9684: [BUG] cannot read rarquet file written with partition_cols
**Describe the bug**
I write a DataFrame to parquet with partition_cols parameter set. When I try to read it back, it can not read it properly. 

**Steps/Code to reproduce bug**
```
>>> import cudf
>>> df=cudf.DataFrame({"a": [1,2,2,1,3], "b": [44,55,66,22,11]})
>>> df
   a   b
0  1  44
1  2  55
2  2  66
3  1  22
4  3  11
>>> df.to_parquet("tmp/prq_df/", partition_cols=["a"])
>>> import os
>>> os.listdir("tmp/prq_df")
['a=1', 'a=2', 'a=3']
>>> df2=cudf.read_parquet("tmp/prq_df")
>>> df2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/cudf/core/dataframe.py", line 1549, in __repr__
    return self._clean_renderable_dataframe(output)
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/cudf/core/dataframe.py", line 1428, in _clean_renderable_dataframe
    output = output.to_pandas().to_string(
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/cudf/core/dataframe.py", line 5680, in to_pandas
    out_data[i] = self._data[col_key].to_pandas(
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/cudf/core/column/numerical.py", line 601, in to_pandas
    pd_series.index = index
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/pandas/core/generic.py", line 5500, in __setattr__
    return object.__setattr__(self, name, value)
  File "pandas/_libs/properties.pyx", line 70, in pandas._libs.properties.AxisProperty.__set__
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/pandas/core/series.py", line 559, in _set_axis
    self._mgr.set_axis(axis, labels)
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 216, in set_axis
    self._validate_set_axis(axis, new_labels)
  File "/scratch_hdd/auyar/gcylon_dev/lib/python3.8/site-packages/pandas/core/internals/base.py", line 57, in _validate_set_axis
    raise ValueError(
ValueError: Length mismatch: Expected axis has 5 elements, new values have 1 elements


```

**Expected behavior**
I should be able to get the dataframe that I have written to the file. 

**Environment overview (please complete the following information)**
 - Environment location: [Bare-metal]
 - Method of cuDF install: [conda]

**Environment details**
cudf 21.10 installed. 

**Additional context**
Nothing more.

### LINKED ISSUE #9690: [BUG] to_parquet partition_cols fails to write to GCP buckets
the following code works:
```
df.to_parquet(f'gs://<bucket>/<path>/', compressed='snappy', 
              storage_options={"token": "/root/connection.json"})
```
but, if i add `partition_cols` parameter to_parquet fails with following error:

```
df.to_parquet(f'gs://<bucket>/<path>/', compressed='snappy', 
                             partition_cols=['a'], partition_file_name=f'a.parquet', 
              storage_options={"token": "/root/connection.json"})
```
error:
```
_request non-retriable exception: Anonymous caller does not have storage.objects.create access to the Google Cloud Storage object., 401
Traceback (most recent call last):
  File "/conda/envs/rapids/lib/python3.7/site-packages/gcsfs/retry.py", line 111, in retry_request
    return await func(*args, **kwargs)
  File "/conda/envs/rapids/lib/python3.7/site-packages/gcsfs/core.py", line 332, in _request
    validate_response(status, contents, path)
  File "/conda/envs/rapids/lib/python3.7/site-packages/gcsfs/retry.py", line 98, in validate_response
    raise HttpError(error)
gcsfs.retry.HttpError: Anonymous caller does not have storage.objects.create access to the Google Cloud Storage object., 401
...
/conda/envs/rapids/lib/python3.7/site-packages/gcsfs/retry.py in validate_response(status, content, path)
     96             raise ValueError("Bad Request: %s\n%s" % (path, msg))
     97         elif error:
---> 98             raise HttpError(error)
     99         elif status:
    100             raise HttpError({"code": status, "message": msg})  # text-like

HttpError: Anonymous caller does not have storage.objects.create access to the Google Cloud Storage object., 401
```
At first it looks like auth error and config hasn't been set right. if it was an issue with authentication, it should have errored in the case of first `to_parquet` without `partition_cols` too. but the write these is successful.

### PR DESCRIPTION
**Pull request by rjzamora**
Closes #9684 
Closes #9690

This PR refactors path handling in `cudf.read_parquet` and uses `pyarrow.dataset` to support for directory-partitioned datasets (with full filterings support at row-group granularity). Since it is my understanding that some users may wish for directory-partitioned columns to be represented as a raw dtype (rather than always becoming categorical), I also added an optional `categorical_partitions` argument (open to suggestions on a better name).

### CODE REVIEW

File: python/cudf/cudf/io/parquet.py [hunk 1]
```diff
@@ -410,6 +502,124 @@ def read_parquet(
...
+                )
+            else:
+                # Not building categorical columns, so
+                # `value` is already what we want
+                dfs[-1][name] = (
+                    as_column(value)
+                    .as_frame()
+                    .repeat(len(dfs[-1]))
+                    ._data[None]
+                )
```
Comment (rjzamora) [inline python/cudf/cudf/io/parquet.py:598]: Here, we convert a scalar value into a uniform column (with that value repeated in every element). I copied this approach from the partitioned-read logic in dask-cudf.  Is this still the preferred way to build such a column? cc @shwina @galipremsagar

Comment (shwina) [reply python/cudf/cudf/io/parquet.py:598]: I think `as_column(scalar, length=s)` is what you want

**Comment by rjzamora** [comment]
rerun tests

**Comment by rjzamora** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]

**Review by quasiben — APPROVED** [review]

**Comment by rjzamora** [comment]
@gpucibot merge

**Pull request merged**